deepchem/models/keras_model.py

```diff
@@ -969,14 +971,89 @@ def restore(self, checkpoint=None, model_dir=None):
     if tf.executing_eagerly():
       self._checkpoint.restore(checkpoint)
     else:
-      self._checkpoint.restore(checkpoint).run_restore_ops(self.session)
+      if session is None:
+        session = self.session
+      self._checkpoint.restore(checkpoint).run_restore_ops(session)
 
   def get_global_step(self):
     """Get the number of steps of fitting that have been performed."""
     if tf.executing_eagerly():
       return int(self._global_step)
     return self._global_step.eval(session=self.session)
 
+  def default_assignment_map(self, source_model, include_top=True, **kwargs):
+    """
+    Creates a default assignment map between layers of source and current model.
+    This is used only when a custom assignment map is missing.
+
+    Parameters
+    ----------
+    source_model: dc.models.KerasModel
+        Source model to copy variable values from.
+    include_top: bool, default True
+        if true, copies the last dense layer
+    """
+    assignment_map = {}
+
+    for idx, layer in enumerate(source_model.model.layers[:-1]):
+      assignment_map[layer] = self.model.layers[idx]
+    if include_top:
+      assignment_map[source_model.model.layers[-1]] = self.model.layers[-1]
+
+    return assignment_map
+
+  def load_from_pretrained(self,
+                           source_model,
+                           assignment_map=None,
+                           checkpoint=None,
+                           model_dir=None,
+                           include_top=True,
+                           **kwargs):
+    """Loads a set of layer weights from a pretrained model. The method takes in
+    a source model and an assignment map, and several other optional arguments.
+    The assignment map is a dictionary between the layers of the source model
+    and the layers in the current model, whose weights we want to copy over. If
+    an assignment map is not supplied, a default assignment map is generated,
+    which assumes the architecture has a Dense layer as its last layer and can be
+    included by setting include_top to True. The default assignment map allows
+    applying an existing trained model to a different multi-task setting involving
+    the same model or to a different task like classification from regression or
+    vice-versa.
+
+    Parameters
+    ----------
+    assignment_map: Dict, default None
+      Dictionary containing layer mapping between source and current model layers
```

Comment:
I think this should be a map between variables, not layers.  Remember, not all models are composed of layers!

Reply:
`KerasModel` can be built in two ways:
1. Using the functional API which is a DAG of layers. This would obviously work with the current setup.

2. Subclassing `tf.keras.Model` and implementing `__call__()`. For the second one, all examples are defined similar to the example below. The layers are defined at some point in the `__init__` or `build()` and are added to the `layers` attribute of the model. 
```
class MLP(tf.keras.Model):

  def __init__(self, **kwargs):
    super(MLP, self).__init__(**kwargs)
    self.dense_1 = layers.Dense(64, activation='relu')
    self.dense_2 = layers.Dense(10)

  def call(self, inputs):
    x = self.dense_1(inputs)
    return self.dense_2(x)

example_model = MLP()
print(example_model.layers)
```
The output has both the dense layers defined. Same thing happens when used with `build`. However if the variables or layers are defined under `call`, then this won't work. But is that usual practice?

I can switch back to variables, but its easier to keep track of layers (and associated variables indirectly) because there are fewer of them when compared to variables.

Reply:
It doesn't even need to use layers at all.  It can define variables directly in `__init__()` or `build()`, then make use of them in `call()`.  This style of model definition was added recently.  It closely matches the way models are defined in PyTorch, so I'd expect it to be pretty common going forward.

Another option is that we could allow the map to contain either layers or variables.  I'm not sure that really adds much, though, other than being a potential source of confusion.